docs/source/tutorial/codec.rst

```diff
@@ -0,0 +1,365 @@
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+   implied.  See the License for the specific language governing
+   permissions and limitations under the License.
+
+
+.. _codec:
+
+#########################
+Codec & HOCON Job Configs
+#########################
+
+
+What is Codec/HOCON?
+====================
+
+`Codec <https://github.com/addthis/codec>`_ is a java serialization library by AddThis that adds more helpful features
+on top of `typesafe config (HOCON) <https://github.com/typesafehub/config>`_ and jackson (JSON). All hydra jobs configs
+are parsed by Codec to turn them into Java objects for running in hydra.
+
+We often refer to our job config language internally as "HOCON", but it is really a specific dialect of HOCON that Codec
+knows how to parse. For example, things like _default, _primary, _inline, plugins, globals, and the like are all
+Codec extensions to HOCON.
+
+Codec is how we load all the things you can use in a job - Bundle Filters, Value Filters, Path elements, etc.
+They are all defined in .conf files in the hydra source code.
+
+
+JSON vs HOCON
+=============
+
+Since July 2014, and hydra version 4.2.14, hydra has had support for some form of HOCON. However, prior to this release
+the only job format supported was the JSON format.
+
+JSON format jobs are actually somewhat of a misnomer. Most JSON jobs that we wrote internally at AddThis were not
+actually "JSON" in the strict sense. They were more of a javascript format, as Codec 2 supported non-quoted
+identifiers and javascript-style comments, both illegal in JSON.
+
+JSON is actually a subset of HOCON, which means that any JSON hydra job is actually also a HOCON job. HOCON also
+supports the non-quoted identifiers and comments that our "JSON" job format supported. So if you were
+to ask "How do I convert a JSON job to HOCON" the answer is actually "Your job is already HOCON"! However, Hydra
+only enables the HOCON parser if a job does not have curly braces at the start and end.
+
+Some key differences:
+---------------------
+* Values in HOCON don't have to be quoted. For most production hydra jobs, our convention is to use quotes anyways.
+* Single quoted strings are not supported in HOCON, while they were in the JSON format. In HOCON, a single quoted
+  string will result in the string including the single quote characters.
+* Comments in HOCON can be made using either "//" or "#". Our JSON format only allowed "//".
+* There is only one additional reserved keyword in HOCON: "include". You must not use "include" anywhere in a job config
+  except in quoted strings, unless using it for its file include function.
+* You must escape strings as in java. In the JSON format, you could do something like "\\" or '"', which would now need
```

Comment:
the "\\" example would parse/ validate, but it was not actually working correctly in almost every case